File: Cpp/Find_the_Index_of_the_First_Occurrence_in_a_String.cpp
class Solution {
public:
    void computeLPS(std::string& needle, std::vector<int>& lps) {
        int m = needle.length();
        lps[0] = 0;
        int length = 0;
        int i = 1;

        while (i < m) {
            if (needle[i] == needle[length]) {
                length++;
                lps[i] = length;
                i++;
            } else {
                if (length != 0) {
                    length = lps[length - 1];
                } else {
                    lps[i] = 0;
                    i++;
                }
            }
        }
    }
    int strStr(string haystack, string needle) {
        if (needle.empty())
            return 0;

        int n = haystack.length();
        int m = needle.length();

        if (m > n)
            return -1;
        vector<int> lps(m, 0);
        computeLPS(needle, lps);

        int i = 0;
        int j = 0;

        while (i < n) {
            if (haystack[i] == needle[j]) {
                i++;
                j++;
            }

            if (j == m) {
                return i - j;
            } else if (i < n && haystack[i] != needle[j]) {
                if (j != 0) {
                    j = lps[j - 1];
                } else {
                    i++;
                }
            }
        }
        return -1;
    }
};

File: Rust/Find_the_Index_of_the_First_Occurrence_in_a_String.rs
impl Solution {
    fn compute_lps(needle: &String) -> Vec<usize> {
        let m = needle.len();
        let mut lps = vec![0; m];
        let mut length = 0;
        let mut i = 1;
        let needle_bytes = needle.as_bytes();

        while i < m {
            if needle_bytes[i] == needle_bytes[length] {
                length += 1;
                lps[i] = length;
                i += 1;
            } else {
                if length != 0 {
                    length = lps[length - 1];
                } else {
                    lps[i] = 0;
                    i += 1;
                }
            }
        }
        lps
    }
    pub fn str_str(haystack: String, needle: String) -> i32 {
        if needle.is_empty() { return 0; }

        let n = haystack.len();
        let m = needle.len();

        if m > n { return -1; }

        let lps = Self::compute_lps(&needle);
        let haystack_bytes = haystack.as_bytes();
        let needle_bytes = needle.as_bytes();
        
        let mut i = 0;
        let mut j = 0;

        while i < n {
            if haystack_bytes[i] == needle_bytes[j] {
                i += 1;
                j += 1;
            }

            if j == m {
                return (i - j) as i32;
            } else if i < n && haystack_bytes[i] != needle_bytes[j] {
                if j != 0 {
                    j = lps[j - 1];
                } else {
                    i += 1;
                }
            }
        }
        -1
    }
}